### PR TITLE
reqs: Update cached_property to functools version

### PR DESCRIPTION
**Pull request by EdCaunt**
`cached_property` has been included in `functools` since Python 3.8. The original `cached_property` module hasn't had an update since. Since Python 3.7 is officially end-of-life, support can be dropped and we can move to using `functools.cached_property` throughout the codebase. This PR updates the requirements and all imports accordingly.

### CODE REVIEW

File: devito/data/decomposition.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from collections.abc import Iterable
 
 import numpy as np
-from cached_property import cached_property
+from functools import cached_property
```
Comment (mloubout) [inline devito/data/decomposition.py:4]: This is now a system import so convention throughout is to group it with system import (first group with collection)

File: .github/workflows/pytest-core-nompi.yml [hunk 2]
```diff
@@ -32,11 +32,11 @@ jobs:
         name: [
            pytest-ubuntu-py311-gcc11-noomp,
            pytest-ubuntu-py38-gcc12-omp,
-           pytest-ubuntu-py37-gcc7-omp,
+           pytest-ubuntu-py38-gcc7-omp,
```
Comment (mloubout) [inline .github/workflows/pytest-core-nompi.yml:35]: Ok, I'll fix the branch rules once approved so it shows as green

File: devito/types/dense.py [hunk 3]
```diff
@@ -243,7 +244,20 @@ def shape_with_halo(self):
         """
         return tuple(j + i + k for i, (j, k) in zip(self.shape, self._size_outhalo))
 
-    _shape_with_outhalo = shape_with_halo
+    @cached_property
+    def _shape_with_outhalo(self):
```
Comment (mloubout) [inline devito/types/dense.py:248]: We gotta find a better way, duplication like that are not good.

Comment (EdCaunt) [reply devito/types/dense.py:248]: Yeah, need to figure that one out. Wanted to get it working first though

Comment (EdCaunt) [reply devito/types/dense.py:237]: Now improved somewhat

Comment (mloubout) [reply devito/types/dense.py:237]: Well no, the docstring is not really the issue, the content of the function is which is twice the same code.

File: .github/workflows/pytest-core-nompi.yml [hunk 4]
```diff
@@ -84,8 +88,13 @@ jobs:
           language: "openmp"
           sympy: "1.9"
 
+<<<<<<< HEAD
```
Comment (mloubout) [inline .github/workflows/pytest-core-nompi.yml:91]: rebase leftover

**Review by mloubout — APPROVED** [review]

**Review by FabioLuporini — APPROVED** [review]

**Pull request merged**